File: Player.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.mycompany.mazegame;

/**
 *
 * @author lsuti
 */
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;
import java.awt.image.BufferedImage;

public class Player extends Sprite {

    private double dx = 0, dy = 0;
    private static final double ACC = 0.5, MAX = 6.0;
    private boolean shielded = false, immune = false, speedBoosted = false;
    private int shieldDur = 0, speedDur = 0, immuneDur = 0;
    private int score = 0, combo = 1, powerUpsCollected = 0;
    private Maze maze;
    private StatsTracker stats;
    private Inventory inv;
    private Achievements ach;

    public Player(GamePanel p, Maze m) {
        super(p.loadImage("/images/avatar.png"), 50, 50);
        this.maze = m;
        this.stats = new StatsTracker();
        this.inv = new Inventory();
        this.ach = new Achievements();

        p.addKeyListener(new KeyAdapter() {
            public void keyPressed(KeyEvent e) {
                switch (e.getKeyCode()) {
                    case KeyEvent.VK_UP:
                        dy = Math.max(dy - ACC, -MAX);
                        break;
                    case KeyEvent.VK_DOWN:
                        dy = Math.min(dy + ACC, MAX);
                        break;
                    case KeyEvent.VK_LEFT:
                        dx = Math.max(dx - ACC, -MAX);
                        break;
                    case KeyEvent.VK_RIGHT:
                        dx = Math.min(dx + ACC, MAX);
                        break;
                }
                m.checkCollisionWithWalls(Player.this);
            }

            public void keyReleased(KeyEvent e) {
                switch (e.getKeyCode()) {
                    case KeyEvent.VK_UP:
                    case KeyEvent.VK_DOWN:
                        dy = 0;
                        break;
                    case KeyEvent.VK_LEFT:
                    case KeyEvent.VK_RIGHT:
                        dx = 0;
                        break;
                }
            }
        });
    }

    @Override
    public void update(GamePanel panel, boolean collisionEnabled) {
        double nx = x + dx, ny = y + dy;
        x = (int) nx;
        y = (int) ny;
        maze.checkCollisionWithWalls(this);

        // durations
        if (shielded && --shieldDur <= 0) {
            deactivateShield();
        }
        if (speedBoosted && --speedDur <= 0) {
            deactivateSpeedBoost(panel);
        }
        if (immune && --immuneDur <= 0) {
            immune = false;
        }

        stats.addDistance((int) Math.hypot(dx, dy));
    }

    public void keyPressed(KeyEvent e) {
    }  // handled via inner adapter

    public void keyReleased(KeyEvent e) {
    }

    // power-up effects
    public void activateShield() {
        shielded = true;
        shieldDur = 300;
    }

    public void deactivateShield() {
        shielded = false;
        stats.addPowerUpActiveTime(shieldDur / 10);
    }

    public boolean isShielded() {
        return shielded;
    }

    public void activateSpeedBoost() {
        speedBoosted = true;
        speedDur = 300;
    }

    public void deactivateSpeedBoost(GamePanel p) {
        speedBoosted = false;
        stats.addPowerUpActiveTime(speedDur / 10);
    }

    public void activateImmunity() {
        immune = true;
        immuneDur = 150;
    }

    public boolean isImmune() {
        return immune;
    }

    // score/combo/inventory
    public void addScore(int v) {
        score += v;
    }

    public int getScore() {
        return score;
    }

    public int getComboMultiplier() {
        return combo;
    }

    public void incrementComboMultiplier() {
        combo++;
    }

    public Inventory getInventory() {
        return inv;
    }

    public StatsTracker getStatsTracker() {
        return stats;
    }

    public Achievements getAchievements() {
        return ach;
    }

    public int getPowerUpsCollected() {
        return powerUpsCollected;
    }
}

File: mist/src/app/GAMES/project/Player.ts
import { Sprite } from './Sprite';
import { GamePanel } from './GamePanel';
import { Inventory } from './Inventory';
import { Achievements } from './Achievements';
import { StatsTracker } from './StatsTracker';

export class Player extends Sprite {
    dx = 0;
    dy = 0;
    private static readonly ACC = 0.5;
    private static readonly MAX = 6.0;
    shielded = false;
    immune = false;
    speedBoosted = false;
    shieldDur = 0;
    immuneDur = 0;
    speedDur = 0;
    score = 0;
    combo = 1;
    powerUpsCollected = 0;

    readonly inv = new Inventory();
    readonly ach = new Achievements();
    readonly stats = new StatsTracker();

    constructor(panel: GamePanel, private maze: any) {
        super(panel.loadImage('/images/avatar.png'), 50, 50);
        panel.canvas.tabIndex = 0;
        panel.canvas.addEventListener('keydown', e => this.handleKeyDown(e));
        panel.canvas.addEventListener('keyup', e => this.handleKeyUp(e));
    }

    handleKeyDown(e: KeyboardEvent) {
        switch (e.key) {
            case 'ArrowUp':
                this.dy = Math.max(this.dy - Player.ACC, -Player.MAX);
                break;
            case 'ArrowDown':
                this.dy = Math.min(this.dy + Player.ACC, Player.MAX);
                break;
            case 'ArrowLeft':
                this.dx = Math.max(this.dx - Player.ACC, -Player.MAX);
                break;
            case 'ArrowRight':
                this.dx = Math.min(this.dx + Player.ACC, Player.MAX);
                break;
        }
        this.maze.checkCollisionWithWalls(this);
    }

    handleKeyUp(e: KeyboardEvent) {
        if (e.key === 'ArrowUp' || e.key === 'ArrowDown') this.dy = 0;
        if (e.key === 'ArrowLeft' || e.key === 'ArrowRight') this.dx = 0;
    }

    update(panel: GamePanel, collisionEnabled: boolean) {
        this.x += this.dx;
        this.y += this.dy;
        this.maze.checkCollisionWithWalls(this);

        // Track distance traveled each tick
        this.stats.addDistance(Math.hypot(this.dx, this.dy));

        if (this.shielded && --this.shieldDur <= 0) this.deactivateShield();
        if (this.speedBoosted && --this.speedDur <= 0)
            this.deactivateSpeedBoost(panel);
        if (this.immune && --this.immuneDur <= 0) this.immune = false;
    }

    activateShield() {
        this.shielded = true;
        this.shieldDur = 300;
    }

    deactivateShield() {
        this.shielded = false;
        this.stats.addPowerUpActiveTime(this.shieldDur / 10);
    }

    isShielded(): boolean {
        return this.shielded;
    }

    activateSpeedBoost() {
        this.speedBoosted = true;
        this.speedDur = 300;
    }

    deactivateSpeedBoost(panel: GamePanel) {
        this.speedBoosted = false;
        this.stats.addPowerUpActiveTime(this.speedDur / 10);
    }

    activateImmunity() {
        this.immune = true;
        this.immuneDur = 150;
    }

    isImmune(): boolean {
        return this.immune;
    }

    getInventory(): Inventory {
        return this.inv;
    }

    getComboMultiplier(): number {
        return this.combo;
    }

    incrementComboMultiplier() {
        this.combo++;
    }

    addScore(v: number) {
        this.score += v;
    }

    useKey(): boolean {
        return this.inv.useItem('Key');
    }
}
